File: native/src/reel_algo.cc
// reel_algo.cc — C++ port of ReelAlgo v2.0 (reel feed ranking).
//
// Pure: JS wrapper fetches velocity + affinity maps and passes them in.
// Fixes baked in:
//  - completion signal now meaningful because trackView (JS) actually persists
//    avgWatchPercentage on the 0..1 scale these thresholds expect.
//  - CREATOR_COLD_START.MIN_QUALITY lowered from 0.3 (=30% engagement, never
//    reachable) to 0.06 to match the engagement-rate scale used elsewhere.
//
// Input JSON:
// { reels:[...], userId, nowMs, followingIds:[...],
//   velocityMap:{reelId:vel}, affinityMap:{authorId:aff},
//   sessionDepth, userAudioPrefs:{audioId:aff},
//   negativeSignals:{skippedCreators:[...],hiddenCategories:[...]} }
// Output JSON: [ rankedReel with _score ]
#include "common.hpp"
#include <map>
#include <set>

namespace pulse {
namespace {

struct Cfg {
  double HALF_LIFE_HOURS=24, MAX_AGE_HOURS=168;
  double W_likes=1.0,W_comments=2.5,W_shares=4.0,W_views=0.1,W_watch=3.0,W_saves=2.0;
  double W_completion=4.0,W_rewatch=5.0,W_skip=-2.0;
  double PERSONALIZATION=0.35,FOLLOW_BOOST=1.5;
  double VELOCITY_WEIGHT=2.0,CREATOR_SCORE_WEIGHT=0.15,VERIFIED_BOOST=1.1;
  double COMP_EXCELLENT=0.8,COMP_GOOD=0.5,COMP_POOR=0.2;
  double REWATCH_MIN=1.5;
  double AUDIO_TRENDING=1.3,AUDIO_AFFINITY=0.5;
  int COLD_MAX_POSTS=5; double COLD_BOOST=1.4, COLD_MIN_QUALITY=0.06; // FIX: was 0.3
  double PACE_EARLY=1.3,PACE_MID=0.9,PACE_LATE=1.2; int PACE_EARLY_T=5,PACE_MID_T=15;
  double NEG_SKIP=0.6,NEG_HIDE=0.2;
};
const Cfg C;

std::string authorId(const Json& reel) {
  if (reel["user"].isObject()) return reel["user"]["_id"].str();
  if (reel["user"].isString()) return reel["user"].str();
  if (reel["author"].isObject()) return reel["author"]["_id"].str();
  return reel["author"].str();
}
const Json& author(const Json& reel) {
  if (reel["user"].isObject()) return reel["user"];
  return reel["author"];
}

double engagementScore(const Json& reel) {
  const Json& s = reel["stats"];
  double score = s["likes"].num(reel["likes"].size()) * C.W_likes
               + s["comments"].num(reel["commentsCount"].num(0)) * C.W_comments
               + s["shares"].num(0) * C.W_shares
               + s["views"].num(0) * C.W_views
               + s["saves"].num(0) * C.W_saves;

  double completion = s["avgWatchPercentage"].num(0);
  if (completion >= C.COMP_EXCELLENT) score += C.W_completion * 1.5;
  else if (completion >= C.COMP_GOOD) score += C.W_completion;
  else if (completion < C.COMP_POOR && s["views"].num(0) > 50) score += C.W_skip;
  else score += completion * C.W_watch;

  double loops = s["avgLoops"].num(s["avgReplayCount"].num(0));
  if (loops >= C.REWATCH_MIN) score += C.W_rewatch * std::min(loops, 3.0);

  return std::max(0.0, score);
}

double freshnessBoost(double createdAt, double now) {
  double h = hoursSince(createdAt, now);
  if (h < 1) return 2.0; if (h < 6) return 1.5; if (h < 24) return 1.2; return 1.0;
}
double creatorScore(const Json& a) {
  if (!a.isObject()) return 0;
  double score = a["isVerified"].boolean(false) ? C.VERIFIED_BOOST : 0;
  double followers = a["stats"]["followers"].num(0);
  if (followers > 0) score += log10p(followers) * 0.5;
  score += a["stats"]["engagementRate"].num(0) * 2;
  return score * C.CREATOR_SCORE_WEIGHT;
}
double audioBoost(const Json& reel, const Json& prefs) {
  std::string audioId = reel["audio"].isObject() ? reel["audio"]["id"].str() : reel["audioId"].str();
  if (audioId.empty()) return 0;
  double boost = 0;
  if (reel["audio"]["isTrending"].boolean(false)) boost += C.AUDIO_TRENDING - 1;
  boost += prefs[audioId].num(0) * C.AUDIO_AFFINITY;
  return boost;
}
double coldStartBoost(const Json& a) {
  if (!a.isObject()) return 0;
  double posts = a["stats"]["posts"].num(a["stats"]["reels"].num(0));
  if (posts >= C.COLD_MAX_POSTS) return 0;
  double er = a["stats"]["engagementRate"].num(0);
  if (er >= C.COLD_MIN_QUALITY) return C.COLD_BOOST - 1;
  return 0.1;
}
std::string category(const Json& reel) {
  std::string c = reel["category"].str(); if (!c.empty()) return c;
  c = reel["vibe"].str(); if (!c.empty()) return c; return "general";
}

struct Scored { Json reel; double score; std::string cat; };

} // namespace

std::string run_reel_rank(const std::string& in) {
  Json input = Json::parse(in);
  double now = input["nowMs"].num(0);
  const Json& reels = input["reels"];
  if (!reels.isArray()) return Json::array().dump();

  std::set<std::string> followingSet, skippedCreators, hiddenCats;
  for (auto& x : input["followingIds"].arr()) followingSet.insert(x.str());
  for (auto& x : input["negativeSignals"]["skippedCreators"].arr()) skippedCreators.insert(x.str());
  for (auto& x : input["negativeSignals"]["hiddenCategories"].arr()) hiddenCats.insert(toLower(x.str()));
  const Json& velocityMap = input["velocityMap"];
  const Json& affinityMap = input["affinityMap"];
  const Json& audioPrefs = input["userAudioPrefs"];
  int sessionDepth = (int)input["sessionDepth"].num(0);

  std::vector<Scored> scored;
  for (auto& reel : reels.arr()) {
    std::string aid = authorId(reel);
    std::string rid = reel["_id"].str();
    const Json& a = author(reel);

    double score = engagementScore(reel);
    score = timeDecay(score, reel["createdAt"].num(now), now, C.HALF_LIFE_HOURS);
    if (hoursSince(reel["createdAt"].num(now), now) > C.MAX_AGE_HOURS) score = engagementScore(reel) * 0.01;
    score *= freshnessBoost(reel["createdAt"].num(now), now);
    score += velocityMap[rid].num(0) * C.VELOCITY_WEIGHT;

    double aff = affinityMap[aid].num(0) * C.PERSONALIZATION;
    if (followingSet.count(aid)) aff *= C.FOLLOW_BOOST;
    score += aff;
    score += creatorScore(a);
    score += audioBoost(reel, audioPrefs);
    score += coldStartBoost(a);

    if (!aid.empty() && skippedCreators.count(aid)) score *= C.NEG_SKIP;
    std::string cat = toLower(category(reel));
    if (!cat.empty() && hiddenCats.count(cat)) score *= C.NEG_HIDE;

    scored.push_back({reel, score, category(reel)});
  }

  std::sort(scored.begin(), scored.end(), [](const Scored& a, const Scored& b){ return a.score > b.score; });

  // Category diversity (no 3 same in a row; cap per batch)
  std::vector<Scored> result, deferred;
  std::map<std::string,int> catCounts;
  for (auto& s : scored) {
    catCounts[s.cat]++;
    std::string prev = result.size()>0 ? result.back().cat : "";
    std::string prev2 = result.size()>1 ? result[result.size()-2].cat : "";
    if (s.cat==prev && s.cat==prev2) deferred.push_back(s);
    else if (catCounts[s.cat] > 4) deferred.push_back(s);
    else result.push_back(s);
  }
  for (auto& s : deferred) {
    bool ins=false;
    for (size_t i=2;i<result.size();++i) if (result[i-1].cat != s.cat) { result.insert(result.begin()+i, s); ins=true; break; }
    if (!ins) result.push_back(s);
  }

  // Session pacing
  Json out = Json::array();
  for (size_t i=0;i<result.size();++i) {
    int pos = sessionDepth + (int)i;
    double m = pos < C.PACE_EARLY_T ? C.PACE_EARLY : pos < C.PACE_MID_T ? C.PACE_MID : C.PACE_LATE;
    Json r = result[i].reel;
    r["_score"] = result[i].score * m;
    out.push_back(r);
  }
  return out.dump();
}

} // namespace pulse

File: src/Algorithms/_fallback/ReelAlgo.js
/**
 * ReelAlgo v2.0 — Advanced Reel Feed Ranking
 *
 * Upgrades:
 *  - Watch-time completion scoring (finish rate > raw views)
 *  - Re-watch detection bonus (looped = high quality)
 *  - Audio/music affinity integration
 *  - Creator cold start handling (new creators get discovery boost)
 *  - Negative signal handling (skip/hide penalties)
 *  - Session-aware pacing for reel feeds
 *  - Category diversity (prevents same-topic saturation)
 *
 * Exports are 100% backward-compatible.
 */

const UserEngagement = require('../../models/UserEngagement');
const Like = require('../../models/Like');

const CONFIG = {
    HALF_LIFE_HOURS: 24,
    MAX_AGE_HOURS: 168,
    WEIGHTS: {
        likes: 1.0, comments: 2.5, shares: 4.0, views: 0.1,
        avgWatchPercentage: 3.0, saves: 2.0,
        // NEW weights
        completionRate: 4.0,     // Full watch = gold signal
        rewatch: 5.0,            // Rewatched = platinum signal
        skipRate: -2.0           // High skip = penalty
    },
    PERSONALIZATION_WEIGHT: 0.35,
    FOLLOW_BOOST: 1.5,
    DIVERSITY_RATE: 0.1,
    DIVERSITY_RECENCY_BOOST: 0.5,
    VELOCITY_WINDOW_HOURS: 1,
    VELOCITY_WEIGHT: 2.0,
    CREATOR_SCORE_WEIGHT: 0.15,
    VERIFIED_BOOST: 1.1,

    // ── NEW v2.0 Config ──
    COMPLETION: {
        EXCELLENT: 0.8,          // 80%+ watched = excellent
        GOOD: 0.5,              // 50%+ = good
        POOR: 0.2               // Below 20% = poor content signal
    },
    REWATCH: {
        MULTIPLIER: 1.5,        // Looped reels get 1.5x
        MIN_LOOPS: 1.5          // Must watch 1.5x duration to count
    },
    AUDIO: {
        TRENDING_BOOST: 1.3,    // Reel uses trending audio
        AFFINITY_WEIGHT: 0.5    // User likes this audio/song
    },
    CREATOR_COLD_START: {
        MAX_POSTS: 5,           // Creator with < 5 posts = cold start
        DISCOVERY_BOOST: 1.4,   // Boost for discovery
        MIN_QUALITY: 0.06       // 6% engagement rate (fix: 0.3 = 30% was unreachable)
    },
    SESSION: {
        EARLY_BOOST: 1.3,
        MID_DIP: 0.9,
        LATE_SPIKE: 1.2,
        EARLY_THRESHOLD: 5,
        MID_THRESHOLD: 15
    },
    NEGATIVE: {
        SKIP_PENALTY: 0.6,      // User skipped similar reels
        HIDE_PENALTY: 0.2       // User explicitly hid this type
    },
    CATEGORY_DIVERSITY: {
        MAX_CONSECUTIVE_SAME: 2,
        MAX_SAME_IN_BATCH: 4
    }
};

// ── Engagement Scoring — Enhanced ──
function calculateEngagementScore(reel) {
    const stats = reel.stats || {};
    let score = 0;

    // Core engagements
    score += (stats.likes || reel.likes?.length || 0) * CONFIG.WEIGHTS.likes;
    score += (stats.comments || reel.commentsCount || 0) * CONFIG.WEIGHTS.comments;
    score += (stats.shares || 0) * CONFIG.WEIGHTS.shares;
    score += (stats.views || 0) * CONFIG.WEIGHTS.views;
    score += (stats.saves || 0) * CONFIG.WEIGHTS.saves;

    // Watch completion (NEW) — the most important signal for reels
    const completion = stats.avgWatchPercentage || 0;
    if (completion >= CONFIG.COMPLETION.EXCELLENT) {
        score += CONFIG.WEIGHTS.completionRate * 1.5;
    } else if (completion >= CONFIG.COMPLETION.GOOD) {
        score += CONFIG.WEIGHTS.completionRate;
    } else if (completion < CONFIG.COMPLETION.POOR && (stats.views || 0) > 50) {
        score += CONFIG.WEIGHTS.skipRate; // Penalty for low completion with enough views
    } else {
        score += completion * CONFIG.WEIGHTS.avgWatchPercentage;
    }

    // Re-watch detection (NEW) — looped reels are gold
    const avgLoops = stats.avgLoops || stats.avgReplayCount || 0;
    if (avgLoops >= CONFIG.REWATCH.MIN_LOOPS) {
        score += CONFIG.WEIGHTS.rewatch * Math.min(avgLoops, 3); // Cap at 3x
    }

    return Math.max(0, score);
}

// ── Velocity ──
async function calculateVelocity(reelId) {
    return (await Like.getLikeVelocity('reel', reelId, CONFIG.VELOCITY_WINDOW_HOURS)) * CONFIG.VELOCITY_WEIGHT;
}

// ── Time Decay ──
function applyTimeDecay(score, createdAt) {
    const hoursAge = (Date.now() - new Date(createdAt).getTime()) / 3600000;
    if (hoursAge > CONFIG.MAX_AGE_HOURS) return score * 0.01;
    return score * Math.pow(0.5, hoursAge / CONFIG.HALF_LIFE_HOURS);
}

function getFreshnessBoost(createdAt) {
    const hrs = (Date.now() - new Date(createdAt).getTime()) / 3600000;
    if (hrs < 1) return 2.0;
    if (hrs < 6) return 1.5;
    if (hrs < 24) return 1.2;
    return 1.0;
}

// ── Personalization — Enhanced ──
function getPersonalizationBoost(userId, authorId, affinityCache, followingSet) {
    if (!userId || !authorId) return 0;
    let boost = (affinityCache?.get(authorId.toString()) || 0) * CONFIG.PERSONALIZATION_WEIGHT;
    if (followingSet?.has(authorId.toString())) boost *= CONFIG.FOLLOW_BOOST;
    return boost;
}

function getCreatorScore(author) {
    if (!author) return 0;
    let score = author.isVerified ? CONFIG.VERIFIED_BOOST : 0;
    const followers = author.stats?.followers || 0;
    if (followers > 0) score += Math.log10(followers + 1) * 0.5;
    score += (author.stats?.engagementRate || 0) * 2;
    return score * CONFIG.CREATOR_SCORE_WEIGHT;
}

// ── Audio Affinity (NEW) ──
function getAudioBoost(reel, userAudioPrefs = {}) {
    let boost = 0;
    const audioId = reel.audio?.id || reel.audioId;
    if (!audioId) return 0;

    // Trending audio bonus
    if (reel.audio?.isTrending) {
        boost += CONFIG.AUDIO.TRENDING_BOOST - 1;
    }

    // User audio affinity
    const affinity = userAudioPrefs[audioId] || 0;
    boost += affinity * CONFIG.AUDIO.AFFINITY_WEIGHT;

    return boost;
}

// ── Creator Cold Start (NEW) ──
function getCreatorColdStartBoost(author) {
    if (!author) return 0;
    const postCount = author.stats?.posts || author.stats?.reels || 0;
    if (postCount >= CONFIG.CREATOR_COLD_START.MAX_POSTS) return 0;

    // New creator with some quality signal
    const engagementRate = author.stats?.engagementRate || 0;
    if (engagementRate >= CONFIG.CREATOR_COLD_START.MIN_QUALITY) {
        return CONFIG.CREATOR_COLD_START.DISCOVERY_BOOST - 1;
    }

    // Even brand new creators get a small discovery boost
    return 0.1;
}

// ── Session Pacing for Reels (NEW) ──
function applyReelSessionPacing(reels, sessionDepth = 0) {
    return reels.map((reel, i) => {
        const pos = sessionDepth + i;
        let m = 1.0;
        if (pos < CONFIG.SESSION.EARLY_THRESHOLD) m = CONFIG.SESSION.EARLY_BOOST;
        else if (pos < CONFIG.SESSION.MID_THRESHOLD) m = CONFIG.SESSION.MID_DIP;
        else m = CONFIG.SESSION.LATE_SPIKE;
        return { ...reel, _score: (reel._score || 0) * m, _sessionPhase: pos < 5 ? 'early' : pos < 15 ? 'mid' : 'late' };
    });
}

// ── Category Diversity (NEW) ──
function enforceCategoryDiversity(reels) {
    if (reels.length < 3) return reels;

    const result = [];
    const deferred = [];
    const categoryCounts = {};

    for (const reel of reels) {
        const category = reel.category || reel.vibe || 'general';
        categoryCounts[category] = (categoryCounts[category] || 0) + 1;

        // Check consecutive same category
        const prevCat = result.length > 0 ? (result[result.length - 1].category || result[result.length - 1].vibe || 'general') : null;
        const prevPrevCat = result.length > 1 ? (result[result.length - 2].category || result[result.length - 2].vibe || 'general') : null;

        if (category === prevCat && category === prevPrevCat) {
            deferred.push(reel);
        } else if (categoryCounts[category] > CONFIG.CATEGORY_DIVERSITY.MAX_SAME_IN_BATCH) {
            deferred.push(reel);
        } else {
            result.push(reel);
        }
    }

    // Re-insert deferred at suitable gaps
    for (const reel of deferred) {
        let inserted = false;
        const cat = reel.category || reel.vibe || 'general';
        for (let i = 2; i < result.length; i++) {
            const prevCat = result[i - 1].category || result[i - 1].vibe || 'general';
            if (prevCat !== cat) { result.splice(i, 0, reel); inserted = true; break; }
        }
        if (!inserted) result.push(reel);
    }

    return result;
}

// ── Diversity Injection ──
function injectDiversity(rankedReels, allReels) {
    if (!allReels || allReels.length === 0) return rankedReels;
    const diversityCount = Math.floor(rankedReels.length * CONFIG.DIVERSITY_RATE);
    if (diversityCount === 0) return rankedReels;

    const rankedIds = new Set(rankedReels.map(r => r._id.toString()));
    const unranked = allReels.filter(r => !rankedIds.has(r._id.toString()));

    const picks = [];
    for (let i = 0; i < Math.min(diversityCount, unranked.length); i++) {
        const idx = Math.floor(Math.random() * unranked.length);
        picks.push(unranked.splice(idx, 1)[0]);
    }

    const result = [...rankedReels];
    picks.forEach(pick => {
        const insertIdx = Math.floor(Math.random() * (result.length - 3)) + 3;
        result.splice(insertIdx, 0, { ...pick, isDiversity: true });
    });
    return result;
}

// ── Main Ranking — Enhanced ──
async function rankReels(reels, userId, options = {}) {
    if (!reels?.length) return [];

    const {
        includeVelocity = true,
        injectDiversityContent = true,
        followingIds = [],
        sessionDepth = 0,           // NEW
        userAudioPrefs = {},        // NEW
        negativeSignals = {}        // NEW
    } = options;

    const authorIds = [...new Set(reels.map(r => (r.user?._id || r.user || r.author?._id || r.author).toString()))];
    let affinityCache = new Map();
    const followingSet = new Set(followingIds.map(String));
    if (userId) affinityCache = await UserEngagement.getBatchAffinities(userId, authorIds);

    // Negative signal sets
    const skippedCreators = new Set((negativeSignals.skippedCreators || []).map(String));
    const hiddenCategories = new Set((negativeSignals.hiddenCategories || []).map(s => s.toLowerCase()));

    // One aggregation for all reels instead of one count query per reel
    let velocityMap = null;
    if (includeVelocity) {
        velocityMap = await Like.getBatchLikeVelocities('reel', reels.map(r => r._id), CONFIG.VELOCITY_WINDOW_HOURS);
    }

    const scored = await Promise.all(reels.map(async reel => {
        const authorId = (reel.user?._id || reel.user || reel.author?._id || reel.author)?.toString();
        const author = reel.user || reel.author || {};

        let score = calculateEngagementScore(reel);
        score = applyTimeDecay(score, reel.createdAt);
        score *= getFreshnessBoost(reel.createdAt);

        if (velocityMap) score += (velocityMap.get(reel._id.toString()) || 0) * CONFIG.VELOCITY_WEIGHT;

        score += getPersonalizationBoost(userId, authorId, affinityCache, followingSet);
        score += getCreatorScore(author);

        // Audio affinity (NEW)
        score += getAudioBoost(reel, userAudioPrefs);

        // Creator cold start (NEW)
        score += getCreatorColdStartBoost(author);

        // Negative signals (NEW)
        if (authorId && skippedCreators.has(authorId)) {
            score *= CONFIG.NEGATIVE.SKIP_PENALTY;
        }
        const reelCat = (reel.category || reel.vibe || '').toLowerCase();
        if (reelCat && hiddenCategories.has(reelCat)) {
            score *= CONFIG.NEGATIVE.HIDE_PENALTY;
        }

        return { ...reel, _score: score, _personalBoost: getPersonalizationBoost(userId, authorId, affinityCache, followingSet) };
    }));

    scored.sort((a, b) => b._score - a._score);

    // Positional passes AFTER the single sort — category diversity defines the
    // final order, so no re-sort afterwards (the old final sort undid it)
    let result = enforceCategoryDiversity(scored);

    // Session pacing — metadata/score tagging on final positions
    result = applyReelSessionPacing(result, sessionDepth);

    // Diversity injection
    if (injectDiversityContent && reels.length > 10) {
        return injectDiversity(result, reels);
    }

    return result;
}

async function getForYouFeed(userId, reels, options = {}) {
    const others = reels.filter(r => (r.user?._id || r.user || r.author?._id || r.author)?.toString() !== userId?.toString());
    const discovery = others.length >= 5 ? others : reels;
    return rankReels(discovery, userId, { ...options, injectDiversityContent: true });
}

async function getFollowingFeed(userId, reels, followingIds) {
    const followingSet = new Set(followingIds.map(String));
    const followed = reels.filter(r => {
        const aid = (r.user?._id || r.user || r.author?._id || r.author)?.toString();
        return followingSet.has(aid) || aid === userId?.toString();
    });
    return rankReels(followed, userId, { includeVelocity: false, injectDiversityContent: false, followingIds });
}

module.exports = {
    calculateEngagementScore, applyTimeDecay, getFreshnessBoost,
    getPersonalizationBoost, getCreatorScore, calculateVelocity,
    rankReels, getForYouFeed, getFollowingFeed, injectDiversity,
    // NEW exports
    getAudioBoost, getCreatorColdStartBoost,
    applyReelSessionPacing, enforceCategoryDiversity,
    CONFIG
};
